### PR TITLE
JDK-8321017: Record in JFR that IEEE rounding mode was corrupted by loading a library

### PR DESCRIPTION
**Pull request by MBaesken**
[JDK-8295159](https://bugs.openjdk.org/browse/JDK-8295159) added some IEEE conformance checks and corrections of the floating point environment on Linux and macOS/BSD, and later some UL logging was added too.
However the information is not added to the JFR events, and this should be enhanced.
The already existing NativeLibraryLoad event can be used for storing the additional information, because the IEEE conformance check and fenv  get/set  is placed in the HS dlopen_helper , where already the NativeLibraryLoad  event objects are created/commited .

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8321017](https://bugs.openjdk.org/browse/JDK-8321017): Record in JFR that IEEE rounding mode was corrupted by loading a library (**Bug** - P4)

### Reviewers
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to [c7e63a27](https://git.openjdk.org/jdk/pull/16903/files/c7e63a27035ce0d62ad98f1989dca309b4485791)
 * [Johannes Bechberger](https://openjdk.org/census#jbechberger) (@parttimenerd - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16903/head:pull/16903` \
`$ git checkout pull/16903`

Update a local copy of the PR: \
`$ git checkout pull/16903` \
`$ git pull https://git.openjdk.org/jdk.git pull/16903/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16903`

View PR using the GUI difftool: \
`$ git pr show -t 16903`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16903.diff">https://git.openjdk.org/jdk/pull/16903.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16903#issuecomment-1833934026)

### CODE REVIEW

**Comment by RealCLanger** [comment]
This does not build on MacOS. Removed from SAP internal test queue.

**Comment by theRealAph** [comment]
I'd be careful about worrying users unnecessarily about this. We fully correct this problem if it happens, so an OpenJDK user is not affected. If we fail to fix the FP environment, the most likely reason is that it was broken before the shared library was loaded.

**Comment by MBaesken** [comment]
> If we fail to fix the FP environment, the most likely reason is that it was broken before the shared library was loaded.

Yes, but in those cases the user should be informed .  The user can then look at the libs on the runtime system and see if there are any candidates causing the issue (like in our RHEL aarch64 example).
There is already the option to trigger an assert  (but few users will run (fast)debug binaries)  and the option to get information from UL (but who will have UL  enabled - for sure not all users).  But quite a few people use JFR so it is a nice way to get the info 'recorded' .

**Comment by egahlin** [comment]
/label jfr

I'm not sure I understand the issue, but adding a field to an event because of GCC bug seems excessive.

**Comment by MBaesken** [comment]
> /label jfr
> 
> I'm not sure I understand the issue, but adding a field to an event because of a GCC bug seems excessive.

I think in practise the issue happens mostly when using binaries (shared libs) compiled by gcc with certain flags.
But it could happen as well with other shared libs loaded by the JVM that mess around with the fp environment ; that's why we check and reset now the fp env.
Maybe Andrew Haley  could comment, he knows better about the details of these issues .

**Comment by MBaesken** [comment]
> I also think it seems odd to include it in JFR.
> 

If you look at the JFR goal  "Provide a low-overhead data collection framework for troubleshooting Java applications and the HotSpot JVM."   (https://openjdk.org/jeps/328)  I think this fits quite well into the "troubleshooting" part.
Not saying that your idea

> As an alternative, I propose to put a test for that condition into the JNI checker (`-Xcheck:jni`). The point of that option is to check third-party native code for problems, so its a perfect fit.

is bad; what do other think ?
My concern here that the JNI checker  (JNI checker (`-Xcheck:jni`)) is probably even less used than JFR. But I might be wrong.
The cases where we directly dlopen/dlsym/fcn-call in the JDK codebase are probably not covered by the JNI checker, right ?

**Comment by dholmes-ora** [comment]
> The cases where we directly dlopen/dlsym/fcn-call in the JDK codebase are probably not covered by the JNI checker, right ?

Right. JNI checking is for checking actual JNI API functions, so I don't see where this would go. A call from Java code into a native method (from which native code could trigger the problem) is not a JNI call.

**Comment by MBaesken** [comment]
> > /label jfr
> > I'm not sure I understand the issue, but adding a field to an event because of a GCC bug seems excessive.
> 
> It's a nasty hard-to-find bug that breaks Java compatibility. Some people have wondered if this is a real-world problem, and the answer is that it's happening, right now, in Oracle's CI testing.

Interesting, do you have some details about the 'Oracle CI testing' occurrence ? If so, what lib caused it ?
Do you think it would be beneficial to have it in the JFR for this particular case  (maybe as a separate event if this is prefered over the current suggestion) ?

**Comment by MBaesken** [comment]
> > The cases where we directly dlopen/dlsym/fcn-call in the JDK codebase are probably not covered by the JNI checker, right ?
> 
> Right. JNI checking is for checking actual JNI API functions, so I don't see where this would go. A call from Java code into a native method (from which native code could trigger the problem) is not a JNI call.

To be fair, those cases where we directly call  dlopen/dlsym/fcn-call in the JDK codebase and trigger the issue,  are not well covered anyway by the current HS coding because they are not going through the os::dll_load, so the check and error message is only observed in the next os::dll_load afterwards. And I think those cases currently cannot be corrected in os::dll_load because the fp env is already 'bad' before os::dll_load .

**Comment by MBaesken** [comment]
the description of ` -Xcheck:jni `  says 
https://docs.oracle.com/javase/8/docs/technotes/guides/troubleshoot/clopts002.html
"The -Xcheck:jni Option
This option is useful in diagnosing problems with applications that use the Java Native Interface (JNI). Sometimes bugs in the native code can cause the HotSpot VM to crash or behave incorrectly."

Not sure if this IEEE conformance issue fits perfectly well into what is said above, but looking at the discussion in this PR, it seems to be controversial .

**Comment by dholmes-ora** [comment]
I had forgotten about the periodic checks related to `-Xcheck:jni`. Yes we probably could do something here.

**Comment by MBaesken** [comment]
> I had forgotten about the periodic checks related to `-Xcheck:jni`. Yes we probably could do something here.

Hi David, what periodic JNI checks are you talking about ?

**Comment by dholmes-ora** [comment]
I previously stated "JNI checking is for checking actual JNI API functions" but Thomas reminded me above that it also enables periodic checks e.g. for the signal handlers, so we already go beyond just checking JNI API function calls. Hence there is scope for performing these FP checks under -Xcheck:jni.

**Comment by theRealAph** [comment]
> > > /label jfr
> > > I'm not sure I understand the issue, but adding a field to an event because of a GCC bug seems excessive.
> > 
> > 
> > It's a nasty hard-to-find bug that breaks Java compatibility. Some people have wondered if this is a real-world problem, and the answer is that it's happening, right now, in Oracle's CI testing.
> 
> Interesting, do you have some details about the 'Oracle CI testing' occurrence ? If so, what lib caused it ?

It was here:
https://bugs.openjdk.org/browse/JDK-8320151

> Do you think it would be beneficial to have it in the JFR for this particular case (maybe as a separate event if this is prefered over the current suggestion) ?

Yes. I think it would be strange to have such a nasty bug and have JFR not record the fact.

**Comment by tstuefe** [comment]
I vote for doing both; the checkjni thing can be done in a separate RFE.

File: src/hotspot/os/bsd/os_bsd.hpp [hunk 1]
```diff
@@ -70,7 +70,7 @@ class os::Bsd {
   // Real-time clock functions
   static void clock_init(void);
 
-  static void *dlopen_helper(const char *path, int mode);
+  static void *dlopen_helper(const char *path, int mode, char *ebuf, int ebuflen);
```
Comment (tstuefe) [inline src/hotspot/os/bsd/os_bsd.hpp:73]: Per-existing, but these helpers should not have been exposed via os::Bsd or os::Linux, they should be filescope static.

File: src/hotspot/os/bsd/os_bsd.cpp [hunk 2]
```diff
@@ -1005,27 +1005,44 @@ void *os::Bsd::dlopen_helper(const char *filename, int mode) {
   assert(rtn == 0, "fegetenv must succeed");
 #endif // IA32
```
Comment (tstuefe) [inline src/hotspot/os/bsd/os_bsd.cpp:1007]: Its difficult to see what exactly changed on MacOS. Is this restructuring necessary?

**Comment by egahlin** [comment]
I like to better understand the problem. Do we expect this to be an issue in 10-15 years, or is it more of a temporary thing? Would we be comfortable adding a method to a MXBean to check if this has happened?

**Comment by theRealAph** [comment]
> I like to better understand the problem. Do we expect this to be an issue in 10-15 years, or is it more of a temporary thing?

it's been a problem for at least 15 years. Rounding modes are an endless game of whack-a-mole.

File: src/hotspot/os/bsd/os_bsd.cpp (hunk 2 above)
Comment (MBaesken) [reply src/hotspot/os/bsd/os_bsd.cpp:1007]: I wanted to bring the placemenet of the JFR event and also UL logging of Linux (dlopen_helper)  and BSD/macOS closer together. With the new structure it is more 'uniform' .

**Comment by egahlin** [comment]
If we believe this to be a reoccurring problem in the foreseeable future, I'm fine with adding the fields. I'm also in favour of adding logging, which I think is better suited for backports than changes event metadata.

**Review by tstuefe — APPROVED** [review]
Looks good to me, provided @egahlin is okay with this.

**Comment by MBaesken** [comment]
Hi Thomas, thanks for the review !

File: src/hotspot/share/jfr/metadata/metadata.xml [hunk 3]
```diff
@@ -957,6 +957,8 @@
     <Field type="string" name="name" label="Name" />
     <Field type="boolean" name="success" label="Success" description="Success or failure of the load operation" />
     <Field type="string" name="errorMessage" label="Error Message" description="In case of a load error, error description" />
+    <Field type="boolean" name="fpEnvCorrectionAttempt" label="FPU Environment correction" description="In case of IEEE conformance issues we might reset the FP environment" />
+    <Field type="boolean" name="fpEnvCorrectionSuccess" label="FPU Environment correction result" description="Stores result in case FP environment correction" />
```
Comment (parttimenerd) [inline src/hotspot/share/jfr/metadata/metadata.xml:961]: minor grammatical issue: "Stores the result in the case of an FP environment correction" is better

**Review by parttimenerd — APPROVED** [review]
Looks good to me

**Comment by MBaesken** [comment]
Hi Johannes, thanks for the review !

**Pull request closed without merging**